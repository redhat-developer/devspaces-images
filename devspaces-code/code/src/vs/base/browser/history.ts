/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { Event } from '../common/event.js';

export interface IHistoryNavigationWidget {

	readonly element: HTMLElement;

	showPreviousValue(): void;

	showNextValue(): void;

	onDidFocus: Event<void>;

	onDidBlur: Event<void>;

}
