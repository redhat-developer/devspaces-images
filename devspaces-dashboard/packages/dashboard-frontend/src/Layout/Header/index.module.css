/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

.headerShow {
  animation-name: showHeader;
  animation-duration: 0.1s;
  animation-timing-function: linear;
}

@keyframes showHeader {
  0% {
    min-height: 0;
    max-height: 0;
    overflow: hidden;
  }

  100% {
    min-height: 57px;
    max-height: 57px;
    overflow: visible;
  }
}

.headerHide {
  min-height: 0;
  max-height: 0;
  overflow: hidden;
  animation-name: hideHeader;
  animation-duration: 0.1s;
  animation-timing-function: linear;
}

@keyframes hideHeader {
  0% {
    min-height: 57px;
    max-height: 57px;
    overflow: visible;
  }

  100% {
    min-height: 0;
    max-height: 0;
    overflow: hidden;
  }
}
