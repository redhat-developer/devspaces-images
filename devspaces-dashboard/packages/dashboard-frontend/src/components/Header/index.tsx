/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import {
  Flex,
  FlexItem,
  PageSection,
  PageSectionVariants,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import WorkspaceStatusLabel from '../WorkspaceStatusLabel';
import {
  DeprecatedWorkspaceStatus,
  DevWorkspaceStatus,
  WorkspaceStatus,
} from '../../services/helpers/types';

const SECTION_THEME = PageSectionVariants.light;

type Props = {
  status: WorkspaceStatus | DevWorkspaceStatus | DeprecatedWorkspaceStatus;
  title: string;
};

class Header extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { title, status } = this.props;

    return (
      <PageSection variant={SECTION_THEME}>
        <Flex>
          <FlexItem alignSelf={{ default: 'alignSelfCenter' }}>
            <TextContent>
              <Text component={TextVariants.h1}>{title}</Text>
            </TextContent>
          </FlexItem>
          <FlexItem>
            <WorkspaceStatusLabel status={status} />
          </FlexItem>
        </Flex>
      </PageSection>
    );
  }
}

export default Header;
