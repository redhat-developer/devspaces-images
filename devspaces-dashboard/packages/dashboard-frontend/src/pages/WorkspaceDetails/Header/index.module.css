/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

.breadcrumb li:first-child {
  margin-right: 0;
}

.breadcrumb li:first-child a {
  margin-right: 8px;
}

.actionButtons > * {
  margin-left: 10px;
}
