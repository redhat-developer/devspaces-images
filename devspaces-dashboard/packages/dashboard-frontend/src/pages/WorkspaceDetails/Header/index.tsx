/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import {
  Flex,
  FlexItem,
  PageSection,
  PageSectionVariants,
  Stack,
  StackItem,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import { Breadcrumb, BreadcrumbItem } from '@patternfly/react-core';
import WorkspaceStatusLabel from '../../../components/WorkspaceStatusLabel';

import styles from './index.module.css';
import {
  DeprecatedWorkspaceStatus,
  DevWorkspaceStatus,
  WorkspaceStatus,
} from '../../../services/helpers/types';

type Props = {
  workspacesLink: string;
  status: WorkspaceStatus | DevWorkspaceStatus | DeprecatedWorkspaceStatus;
  workspaceName: string;
};

class Header extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { workspaceName, status, children, workspacesLink } = this.props;

    return (
      <PageSection variant={PageSectionVariants.light}>
        <Stack hasGutter={true}>
          <StackItem>
            <Breadcrumb className={styles.breadcrumb}>
              <BreadcrumbItem to={workspacesLink}>Workspaces</BreadcrumbItem>
              <BreadcrumbItem isActive>{workspaceName}</BreadcrumbItem>
            </Breadcrumb>
          </StackItem>
          <StackItem>
            <Flex>
              <FlexItem alignSelf={{ default: 'alignSelfCenter' }}>
                <TextContent>
                  <Text component={TextVariants.h1}>{workspaceName}</Text>
                </TextContent>
              </FlexItem>
              <FlexItem>
                <WorkspaceStatusLabel status={status} />
              </FlexItem>
              <FlexItem className={styles.actionButtons} align={{ default: 'alignRight' }}>
                {children}
              </FlexItem>
            </Flex>
          </StackItem>
        </Stack>
      </PageSection>
    );
  }
}

export default Header;
