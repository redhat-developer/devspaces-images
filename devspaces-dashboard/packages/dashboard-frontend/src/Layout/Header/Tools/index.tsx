/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { History } from 'history';
import React from 'react';
import gravatarUrl from 'gravatar-url';
import {
  Avatar,
  PageHeaderTools,
  PageHeaderToolsGroup,
  PageHeaderToolsItem,
} from '@patternfly/react-core';
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../../store';
import { selectBranding } from '../../../store/Branding/selectors';
import { selectUserProfile } from '../../../store/UserProfile/selectors';
import { AboutMenu } from './AboutMenu';
import UserMenu from './UserMenu';
import { ApplicationsMenu } from './ApplicationsMenu';
import { selectApplications } from '../../../store/ClusterInfo/selectors';

type Props = MappedProps & {
  history: History;
  user: che.User | undefined;
  logout: () => void;
};
export class HeaderTools extends React.PureComponent<Props> {
  constructor(props: Props) {
    super(props);
  }

  public render(): React.ReactElement {
    const { applications, userProfile } = this.props;
    const userEmail = userProfile.email;
    const imageUrl = userEmail ? gravatarUrl(userEmail, { default: 'retro' }) : '';
    const isUserAuthenticated = !!userEmail;

    return (
      <>
        <PageHeaderTools>
          <PageHeaderToolsGroup>
            {applications.length !== 0 && <ApplicationsMenu applications={applications} />}
            <PageHeaderToolsItem>
              <AboutMenu
                branding={this.props.branding}
                user={this.props.user}
                userProfile={this.props.userProfile}
              />
            </PageHeaderToolsItem>
            {isUserAuthenticated && (
              <PageHeaderToolsItem>
                <UserMenu
                  branding={this.props.branding}
                  history={this.props.history}
                  user={this.props.user}
                  userProfile={this.props.userProfile}
                  logout={() => this.props.logout()}
                />
              </PageHeaderToolsItem>
            )}
          </PageHeaderToolsGroup>
          {isUserAuthenticated && <Avatar src={imageUrl} alt="Avatar image" />}
        </PageHeaderTools>
      </>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  userProfile: selectUserProfile(state),
  branding: selectBranding(state),
  applications: selectApplications(state),
});

const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(HeaderTools);
