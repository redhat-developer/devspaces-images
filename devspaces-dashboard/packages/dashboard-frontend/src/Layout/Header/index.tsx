/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { History } from 'history';
import React from 'react';
import { Brand, PageHeader } from '@patternfly/react-core';
import { User } from 'che';

import HeaderTools from './Tools';
import { ThemeVariant } from '../themeVariant';

import * as styles from './index.module.css';

type Props = {
  history: History;
  isVisible: boolean;
  logoUrl: string;
  user: User | undefined;
  logout: () => void;
  toggleNav: () => void;
  changeTheme: (theme: ThemeVariant) => void;
};
type State = {
  isVisible: boolean;
};

export default class Header extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isVisible: this.props.isVisible,
    };
  }

  private toggleNav(): void {
    this.props.toggleNav();
  }

  public componentDidUpdate(prevProps: Props): void {
    if (prevProps.isVisible !== this.props.isVisible) {
      this.setState({
        isVisible: this.props.isVisible,
      });
    }
  }

  public render(): React.ReactElement {
    const logo = <Brand src={this.props.logoUrl} alt="Logo" />;

    const className = this.state.isVisible ? styles.headerShow : styles.headerHide;

    return (
      <PageHeader
        style={{ zIndex: 'inherit' }}
        className={className}
        logoComponent="div"
        logo={logo}
        showNavToggle={true}
        onNavToggle={() => this.toggleNav()}
        headerTools={
          <HeaderTools
            history={this.props.history}
            user={this.props.user}
            logout={() => this.props.logout()}
          />
        }
      />
    );
  }
}
