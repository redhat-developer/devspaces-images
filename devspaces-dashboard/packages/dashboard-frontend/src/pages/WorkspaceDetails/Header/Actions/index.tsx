/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { AlertVariant } from '@patternfly/react-core';
import React from 'react';
import { History } from 'history';
import common from '@eclipse-che/common';
import WorkspaceActionsProvider from '../../../../containers/WorkspaceActions';
import {
  WorkspaceAction,
  WorkspaceStatus,
  DevWorkspaceStatus,
  DeprecatedWorkspaceStatus,
} from '../../../../services/helpers/types';
import {
  ActionContextType,
  WorkspaceActionsConsumer,
} from '../../../../containers/WorkspaceActions/context';
import { lazyInject } from '../../../../inversify.config';
import { AppAlerts } from '../../../../services/alerts/appAlerts';
import getRandomString from '../../../../services/helpers/random';
import DropdownActions from './Dropdown';
import ButtonAction from './Button';

type Props = {
  workspaceId: string;
  workspaceName: string;
  canDelete: boolean;
  status: WorkspaceStatus | DevWorkspaceStatus | DeprecatedWorkspaceStatus;
  history: History;
};

export class HeaderActionSelect extends React.PureComponent<Props> {
  @lazyInject(AppAlerts)
  private appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);
  }

  private async handleSelectedAction(
    selectedAction: WorkspaceAction,
    context: ActionContextType,
  ): Promise<void> {
    try {
      if (selectedAction === WorkspaceAction.DELETE_WORKSPACE) {
        try {
          await context.showConfirmation([this.props.workspaceName]);
        } catch (e) {
          return;
        }
      }
      const nextPath = await context.handleAction(selectedAction, this.props.workspaceId);
      if (!nextPath) {
        return;
      }
      this.props.history.push(nextPath);
    } catch (e) {
      const errorMessage = common.helpers.errors.getMessage(e);
      this.showAlert(errorMessage);
      console.warn(errorMessage);
    }
  }

  private showAlert(message: string): void {
    this.appAlerts.showAlert({
      key: 'workspace-details-' + getRandomString(4),
      title: message,
      variant: AlertVariant.warning,
    });
  }

  render(): React.ReactNode {
    const { canDelete, history, status } = this.props;

    return (
      <WorkspaceActionsProvider history={history}>
        <WorkspaceActionsConsumer>
          {context => {
            if (status === 'Deprecated') {
              if (canDelete) {
                return (
                  <ButtonAction
                    context={context}
                    onAction={(action, context) => this.handleSelectedAction(action, context)}
                  />
                );
              } else {
                return <></>;
              }
            }
            const { ...props } = this.props;
            return (
              <DropdownActions
                {...props}
                context={context}
                onAction={(action, context) => this.handleSelectedAction(action, context)}
              />
            );
          }}
        </WorkspaceActionsConsumer>
      </WorkspaceActionsProvider>
    );
  }
}
