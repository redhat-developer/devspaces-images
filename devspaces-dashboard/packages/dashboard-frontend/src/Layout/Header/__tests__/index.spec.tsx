/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { createHashHistory } from 'history';
import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import Header from '..';

jest.mock('../Tools', () => {
  const FakeTools = (props: { logout: () => void; changeTheme: () => void }) => (
    <React.Fragment>
      <button onClick={() => props.logout()}>logout</button>
      <button onClick={() => props.changeTheme()}>change theme</button>
    </React.Fragment>
  );
  FakeTools.displayName = 'Tools';
  return FakeTools;
});

describe('Page header', () => {
  const mockLogout = jest.fn();
  const mockToggleNav = jest.fn();
  const mockChangeTheme = jest.fn();

  const user = {
    email: 'johndoe@example.com',
    name: 'John Doe',
  } as che.User;
  const logoUrl = 'branding/logo';
  const isHeaderVisible = true;
  const history = createHashHistory();

  const component = (
    <Header
      history={history}
      isVisible={isHeaderVisible}
      logoUrl={logoUrl}
      user={user}
      logout={mockLogout}
      toggleNav={mockToggleNav}
      changeTheme={mockChangeTheme}
    />
  );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the component', () => {
    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });

  it('should fire logout event', () => {
    render(component);

    const logoutButton = screen.getByRole('button', { name: 'logout' });
    fireEvent.click(logoutButton);

    expect(mockLogout).toHaveBeenCalled();
  });
});
