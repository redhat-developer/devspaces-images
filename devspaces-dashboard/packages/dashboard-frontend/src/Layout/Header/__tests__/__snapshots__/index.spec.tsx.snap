// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page header should correctly render the component 1`] = `
<header
  className="pf-c-page__header"
  style={
    Object {
      "zIndex": "inherit",
    }
  }
>
  <div
    className="pf-c-page__header-brand"
  >
    <div
      className="pf-c-page__header-brand-toggle"
    >
      <button
        aria-controls={null}
        aria-disabled={false}
        aria-expanded="true"
        aria-label="Global navigation"
        className="pf-c-button pf-m-plain"
        data-ouia-component-id="OUIA-Generated-Button-plain-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        id="nav-toggle"
        onClick={[Function]}
        role={null}
        type="button"
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          fill="currentColor"
          height="1em"
          role="img"
          style={
            Object {
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 448 512"
          width="1em"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          />
        </svg>
      </button>
    </div>
    <div
      className="pf-c-page__header-brand-link"
    >
      <img
        alt="Logo"
        className="pf-c-brand"
        src="branding/logo"
      />
    </div>
  </div>
  <button
    onClick={[Function]}
  >
    logout
  </button>
  <button
    onClick={[Function]}
  >
    change theme
  </button>
</header>
`;
